File: Wildcard Matching.py
#https://leetcode.com/problems/wildcard-matching/description/

class Solution:
    def isMatch(self, text: str, pattern: str) -> bool:
        dp=[[False for _ in range(len(pattern)+1)]for _ in range(len(text)+1)]
        starFirst=0
        for i in range(len(text)+1):
            for j in range(len(pattern)+1):
                if i==0:
                    if j==0:
                        dp[i][j]=True
                    elif pattern[j-1]=='*' and starFirst!=1:
                        dp[i][j]=True
                    else:
                        dp[i][j]=False
                        starFirst=1
                elif j>0:
                    if text[i-1]==pattern[j-1] or pattern[j-1]=='?':
                        dp[i][j]=dp[i-1][j-1]
                    elif pattern[j-1]=='*':
                        dp[i][j]=max(dp[i-1][j], dp[i][j-1])
        return dp[len(text)][len(pattern)]
        

File: Wildcard Matching.cpp
//https://leetcode.com/problems/wildcard-matching/description/


class Solution {
public:
    bool helper(string &s, string &p, int i, int j, vector<vector<int>> &dp) {

        // ---- base cases ----
        if (i == s.size() && j == p.size())
            return true;

        if (j == p.size())
            return false;

        if (i == s.size()) {
            // remaining pattern must be all '*'
            while (j < p.size() && p[j] == '*') j++;
            return j == p.size();
        }
        if(dp[i][j]!=-1) return dp[i][j];

        // ---- recursive cases ----
        if (p[j] == '*') {
            // not use '*' OR use it
            return dp[i][j]=(helper(s, p, i, j + 1,dp) || helper(s, p, i + 1, j,dp));
        }

        if (p[j] == '?' || p[j] == s[i]) {
            return dp[i][j]=helper(s, p, i + 1, j + 1,dp);
        }

        return false;
    }

    bool isMatch(string s, string p) {
        vector<vector<int>> dp(s.size(),vector<int> (p.size(),-1));
        return helper(s, p, 0, 0, dp);
    }
};
